File: src/data/human/Longest_Valid_Parentheses/Longest_Valid_Parentheses.java
import java.util.Stack;

public class Solution {
    public int longestValidParentheses(String s) {
        Stack<Integer> stack = new Stack<Integer>();
        int max=0;
        int left = -1;
        for(int j=0;j<s.length();j++){
            if(s.charAt(j)=='(') stack.push(j);            
            else {
                if (stack.isEmpty()) left=j;
                else{
                    stack.pop();
                    if(stack.isEmpty()) max=Math.max(max,j-left);
                    else max=Math.max(max,j-stack.peek());
                }
            }
        }
        return max;
    }
}

public class Longest_Valid_Parentheses {
public static void runTests(Solution solution) {
    // Test case 1: Simple valid parentheses
    String s = "(()";
    int expected_result = 2;
    assert solution.longestValidParentheses(s) == expected_result : "Test case 1 failed. Expected " + expected_result;

    // Test case 2: All parentheses are valid
    s = ")()())";
    expected_result = 4;
    assert solution.longestValidParentheses(s) == expected_result : "Test case 2 failed. Expected " + expected_result;

    // Test case 3: No valid parentheses
    s = "(((";
    expected_result = 0;
    assert solution.longestValidParentheses(s) == expected_result : "Test case 3 failed. Expected " + expected_result;

    // Test case 4: Entire string is valid parentheses
    s = "()()()";
    expected_result = 6;
    assert solution.longestValidParentheses(s) == expected_result : "Test case 4 failed. Expected " + expected_result;

    // Test case 5: Nested valid parentheses
    s = "((()))";
    expected_result = 6;
    assert solution.longestValidParentheses(s) == expected_result : "Test case 5 failed. Expected " + expected_result;

    // Test case 6: Mixed valid and invalid parentheses
    s = "(()))())(";
    expected_result = 4;
    assert solution.longestValidParentheses(s) == expected_result : "Test case 6 failed. Expected " + expected_result;

    // Test case 7: Empty string
    s = "";
    expected_result = 0;
    assert solution.longestValidParentheses(s) == expected_result : "Test case 7 failed. Expected " + expected_result;
}

    public static void main(String[] args) {
        Solution solution = new Solution();
        for (int i = 0; i < 200000000; i++) {
            runTests(solution); // Ensure your test cases are implemented inside runTests
        }
    }
}

    

File: src/data/4o/Longest_Valid_Parentheses/Longest_Valid_Parentheses.cpp
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    int longestValidParentheses(string s) {
        int maxLength = 0;
        stack<int> stk;
        stk.push(-1);  // Base index for calculating the valid lengths

        for (int i = 0; i < s.length(); i++) {
            if (s[i] == '(') {
                stk.push(i);
            } else {
                stk.pop();
                if (stk.empty()) {
                    stk.push(i);  // Push the index of the last invalid ')'
                } else {
                    maxLength = max(maxLength, i - stk.top());
                }
            }
        }

        return maxLength;
    }
};

//  Test cases for the C++ solution
#include <cassert>

// Test cases for the C++ solution
void testSolution(Solution& solution) {
    // Test case 1: Simple valid parentheses
    std::string s = "(()";
    int expected_result = 2;
    assert(solution.longestValidParentheses(s) == expected_result && "Test case 1 failed.");

    // Test case 2: All parentheses are valid
    s = ")()())";
    expected_result = 4;
    assert(solution.longestValidParentheses(s) == expected_result && "Test case 2 failed.");

    // Test case 3: No valid parentheses
    s = "(((";
    expected_result = 0;
    assert(solution.longestValidParentheses(s) == expected_result && "Test case 3 failed.");

    // Test case 4: Entire string is valid parentheses
    s = "()()()";
    expected_result = 6;
    assert(solution.longestValidParentheses(s) == expected_result && "Test case 4 failed.");

    // Test case 5: Nested valid parentheses
    s = "((()))";
    expected_result = 6;
    assert(solution.longestValidParentheses(s) == expected_result && "Test case 5 failed.");

    // Test case 6: Mixed valid and invalid parentheses
    s = "(()))())(";
    expected_result = 4;
    assert(solution.longestValidParentheses(s) == expected_result && "Test case 6 failed.");

    // Test case 7: Empty string
    s = "";
    expected_result = 0;
    assert(solution.longestValidParentheses(s) == expected_result && "Test case 7 failed.");
}


int main() {
    Solution solution;
    for (int i = 0; i < 2621440; ++i) {
        testSolution(solution);  
    }
    return 0;
}
